python/astshim/fitsChanContinued.py

```diff
@@ -0,0 +1,258 @@
+from .fitsChan import FitsChan, CardType
+from .fitsChan import *  # noqa: F403 F401
+
+
+def _calc_card_pos(self, index):
+    """Convert a python index into a FitsChan position.
+
+    Parameters
+    ----------
+    self : `FitsChan`
+        The FitsChan to index.
+    index : `int`
+        0-based index into header. If negative, counts from end.
+    """
+    # Calculate 0-based index
+    if index < 0:
+        index = len(self) + index
+
+    # Convert to 1-based index
+    return index + 1
+
+
+def length(self):
+    return self.nCard
+
+
+FitsChan.__len__ = length
+
+
+def iter(self):
+    """The FitsChan itself is the iterator.
+    """
+    self.clearCard()
+    return self
+
+
+FitsChan.__iter__ = iter
+
+
+def next(self):
+    """Return the entire header card until we run out of cards.
+    """
+    card = self.findFits("%f", True)
+    if not card.found:
+        raise StopIteration
+    return card.value
+
+
+FitsChan.__next__ = next
+
+
+def to_string(self):
+    """A FitsChan string representation is a FITS header with newlines
+    after each header card.
+    """
+    return "\n".join(c for c in self)
+
+
+FitsChan.__str__ = to_string
+
+
+def contains(self, name):
+    """Returns True if either the supplied name is present in the FitsChan
+    or the supplied integer is acceptable to the FitsChan.
+    """
+    if isinstance(name, int):
+        # index will be zero-based
+        if name >= 0 and name < self.nCard:
+            return True
+    elif isinstance(name, str):
+        currentCard = self.getCard()
+        try:
+            self.clearCard()
+            result = self.findFits(name, False)
+        finally:
+            self.setCard(currentCard)
+        if result.found:
+            return True
```

Comment:
I disagree. Contains is meant to be telling you if the thing is in the container. It should not raise if the thing is not in the container, it should return False, which it does do.

Reply:
An interesting point. On the one hand for real dicts there is no such exception, but for real dicts the key can be any immutable type so that makes sense. From my perspective Python generally warns you when you try to use the wrong type so I think a TypeError would make a lot of sense here. But I can see your point and will not insist.